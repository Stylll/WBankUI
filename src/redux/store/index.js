import { createStore, applyMiddleware } from 'redux';
import { composeWithDevTools } from 'redux-devtools-extension';
import createSagaMiddleware from 'redux-saga';
import axios from 'axios';

import rootReducer from '../reducers';
import rootSaga from '../sagas';
import { jwtKey } from '../../helpers/defaults';

const sagaMiddleware = createSagaMiddleware();
// const reduxRouterMiddleware = routerMiddleware(history);
const middleware = composeWithDevTools(applyMiddleware(sagaMiddleware));
const store = createStore(rootReducer, middleware);

// eslint-disable-next-line no-undef
const token = localStorage.getItem(jwtKey);

axios.defaults.headers.common.authorization = token;

sagaMiddleware.run(rootSaga);

export default store;
