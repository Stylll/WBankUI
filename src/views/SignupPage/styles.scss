
@import '../../partials/colors';

.container.authpage {
    display: flex;
    justify-content: center;
    text-align: center;

    .signup-container {
        box-shadow: 0 8px 16px 0 rgba(0,0,0,0.2);
        margin: 40px;
        padding: 40px;

        .link {
            font-weight: 100;
            &:hover {
                text-decoration: underline;
                cursor: pointer;
            }
        }
        .error-text {
            color: $red;
        }
    }
}
