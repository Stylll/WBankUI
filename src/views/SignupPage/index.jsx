import React, { useState, useEffect } from 'react';
import { connect } from 'react-redux';
import { useHistory } from 'react-router-dom';
import PropTypes from 'prop-types';
import Header from '../../components/Header/Header.component';
import TextInput from '../../components/TextInput/TextInput.component';
import Button from '../../components/Button/Button.component';
import { validateSignup } from '../../helpers/validators/authValidator';
import * as userActions from '../../redux/actionCreators/userActions';
import './styles.scss';

const SignupPage = ({
  signupUser, isLoading: reduxIsLoading, userErrorMessage,
}) => {
  const [values, setValues] = useState({
    fullname: '',
    email: '',
  });

  const [formErrors, setFormErrors] = useState({
    fullname: '',
    email: '',
  });

  const [isLoading, setIsLoading] = useState(false);

  const history = useHistory();

  useEffect(() => {
    if (!reduxIsLoading) {
      setIsLoading(false);
    }
  }, [reduxIsLoading]);

  const handleChange = (event) => {
    event.preventDefault();
    setValues({
      ...values,
      [event.target.name]: event.target.value,
    });
  };

  const handleClick = (event) => {
    event.preventDefault();
    const { isValid, errors } = validateSignup(values);
    setFormErrors({
      ...errors,
    });
    if (!isValid) {
      return;
    }

    setIsLoading(true);

    signupUser({
      name: values.fullname,
      email: values.email,
    });
  };

  const navigateToSignin = () => {
    history.push('/signin');
  };

  const renderSignupForm = (errorMessage) => (
        <div className="signup-container">
            <div>
                <p>Create an account</p>
                <span className="error-text">{errorMessage}</span>
                <TextInput
                    name="fullname"
                    placeholder="Jim Halpert"
                    title="Enter your fullname"
                    value={values.fullname}
                    error={formErrors.fullname}
                    onChange={handleChange} />
                <TextInput
                    name="email"
                    placeholder="jimhalpert@dundermuflin.com"
                    title="Enter your email"
                    value={values.email}
                    error={formErrors.email}
                    onChange={handleChange} />
                <Button
                    title="Signup"
                    handleClick={handleClick}
                    showLoader={isLoading}
                    disabled={isLoading}
                />
            </div>
            <br />
            <h4 className="link" onClick={() => { navigateToSignin(); }}>
              Already have an account?. click here to login
            </h4>
        </div>
  );

  return (
        <div className="">
            <Header />
            <div className="authpage container">
                <div>
                    <h2 className="intro-text">Welcome To WBank</h2>
                    {renderSignupForm(userErrorMessage)}
                </div>
            </div>
        </div>
  );
};

SignupPage.propTypes = {
  signupUser: PropTypes.func.isRequired,
  isLoading: PropTypes.bool.isRequired,
  userErrorMessage: PropTypes.string.isRequired,
};

const mapStateToProps = ({ user }) => ({
  isLoading: user.isLoading,
  userErrorMessage: user.errorMessage,
});

const mapDispatchToProps = {
  signupUser: userActions.signupUser,
};

export default connect(mapStateToProps, mapDispatchToProps)(SignupPage);
