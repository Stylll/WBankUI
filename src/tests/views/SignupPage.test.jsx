import { render, fireEvent } from '@testing-library/react';
import { Provider } from 'react-redux';
import React from 'react';

import SignupPage from '../../views/SignupPage/index';
import store from '../../redux/store/index';

describe('SignupPage Test', () => {
  it('should render properly', () => {
    const { baseElement } = render(
        <Provider store={store}>
            <SignupPage />
        </Provider>,
    );
    expect(baseElement).toMatchSnapshot();
  });

  it('should validate signup form', () => {
    const { baseElement, getByTestId, getByText } = render(
        <Provider store={store}>
            <SignupPage />
        </Provider>,
    );
    fireEvent.click(getByTestId('signup-signup'));
    expect(getByText('fullname cannot be empty')).toBeInTheDocument();
    expect(getByText('email cannot be empty')).toBeInTheDocument();
  });

  it('should require valid email', () => {
    const { baseElement, getByTestId, getByText } = render(
        <Provider store={store}>
            <SignupPage />
        </Provider>,
    );
    fireEvent.change(getByTestId('signup-email'),
      { target: { value: 'email', name: 'email' } });
    fireEvent.click(getByTestId('signup-signup'));
    expect(getByText('email is invalid')).toBeInTheDocument();
  });

  it('should update the isloading state', () => {
    const { baseElement, getByTestId } = render(
        <Provider store={store}>
            <SignupPage />
        </Provider>,
    );
    fireEvent.change(getByTestId('signup-fullname'),
      { target: { value: 'customer name', name: 'fullname' } });
    fireEvent.change(getByTestId('signup-email'),
      { target: { value: 'email@email.com', name: 'email' } });
    fireEvent.click(getByTestId('signup-signup'));
    expect(store.getState().user.isLoading).toBe(true);
  });
});
