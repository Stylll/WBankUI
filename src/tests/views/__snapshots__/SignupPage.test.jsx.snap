// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SignupPage Test should render properly 1`] = `
<body>
  <div>
    <div
      class=""
    >
      <div
        class="header container"
      >
        <div
          class="link"
        >
          WBank
        </div>
        <div
          class="topRight"
        />
      </div>
      <div
        class="bottom container"
      />
      <div
        class="authpage container"
      >
        <div
          class="form-container"
        >
          <h2
            class="intro-text"
          >
            Welcome To WBank
          </h2>
          <div
            class="signup-container"
          >
            <div>
              <p>
                Create an account
              </p>
              <span
                class="error-text"
              />
              <div
                class="form-input"
              >
                <div
                  class="title"
                >
                  Enter your fullname
                </div>
                <input
                  class="text-input"
                  data-testid="signup-fullname"
                  id=""
                  name="fullname"
                  placeholder="Jim Halpert"
                  type="text"
                  value=""
                />
                <span
                  class="info"
                />
                
              </div>
              <div
                class="form-input"
              >
                <div
                  class="title"
                >
                  Enter your email
                </div>
                <input
                  class="text-input"
                  data-testid="signup-email"
                  id=""
                  name="email"
                  placeholder="jimhalpert@dundermuflin.com"
                  type="text"
                  value=""
                />
                <span
                  class="info"
                />
                
              </div>
              <div
                class="btn-button"
              >
                <button
                  class="button "
                  data-testid="signup-signup"
                  type="submit"
                >
                  Signup
                </button>
              </div>
            </div>
            <br />
            <h4
              class="link"
            >
              Already have an account?. click here to signin
            </h4>
          </div>
        </div>
      </div>
    </div>
  </div>
</body>
`;
